File: 002/ruby/wordsearch.rb
def search(grid, rc, word, path)
  return path if word == ""
  return [] if path.include?(rc) || !["?", grid[rc]].include?(word[0])
  [*-1..1].product([*-1..1]).flat_map {|r,c|
    search(grid, [rc[0]+r,rc[1]+c], word[1..-1], path + [rc])
  }
end

grid_s, word_s = ARGF.read.upcase.split(/\n\n/)
gridline = grid_s.split(/\n/).map.with_index {|line,row|
  line.chars.map.with_index {|ch,col| [[row,col],ch] }
}
grid = Hash[*gridline.flatten(2)]

mark = word_s.split(/,/).flat_map {|word|
  grid.keys.flat_map {|rc| search(grid, rc, word.strip, [])}
}

puts [gridline.map {|rowline|
  rowline.map {|rc, ch| mark.include?(rc) ? ch : '.' }.join(" ")
}.join("\n"), nil, word_s]

File: 002/scala/WordSearch.scala
object WordSearch {
  def search(grid: Map[(Int,Int),Char], rc: (Int,Int), word: String, path: List[(Int,Int)]):List[(Int,Int)] = {
    if (word == "") return path
    if (!grid.keys.toList.contains(rc) || path.contains(rc) || !Array('?',grid(rc)).contains(word.head))
      return Nil
    (for (r <- -1 to 1; c <- -1 to 1) yield (r,c)).toList.flatMap {
      case (r,c) => search(grid, (rc._1+r,rc._2+c), word.tail, rc :: path)
    }
  }
  def main(args: Array[String]) {
    val Array(grid_s, word_s) = scala.io.Source.fromFile(args(0)).mkString.toUpperCase.split("\n\n")
    val gridline = grid_s.split("\n").zipWithIndex.map {
      case (line, row) => line.zipWithIndex.map {
        case (ch, col) => ((row,col), ch)
      }
    }

    val grid = gridline.flatten.toMap
    val mark = word_s.split(",").flatMap {
      case word => grid.keys.flatMap { case rc => search(grid, rc, word.trim, Nil) }.toList
    }

    println(gridline.map {
      case gridline => gridline.map {
        case (rc, ch) => if (mark.contains(rc)) ch else '.'
      }.mkString(" ")
    }.mkString("\n"))
  }
}
